c_binding/libsg.c

```diff
@@ -940,3 +1035,385 @@ int _sg_host_no(char *err_msg, int fd, unsigned int *host_no)
+                                    _LSM_ERR_MSG_LEN),
+                         sense_err_msg);
+        goto out;
+    }
+
+    memcpy(returned_sense_data, sense_data, _T10_SPC_SENSE_DATA_MAX_LENGTH);
+
+out:
+
+    return rc;
+}
+
+int32_t _sg_smart_interpret_asc(uint8_t asc)
+{
+    if (asc == _T10_SPC_ASC_IMPENDING_FAILURE)
+        return LSM_DISK_SMART_STATUS_FAIL;
+    else if (asc == _T10_SPC_ASC_WARNING)
+        return LSM_DISK_SMART_STATUS_WARN;
+    else
+        return LSM_DISK_SMART_STATUS_GOOD;
+}
+
+/*
+ * Query a SAS drive to get its SMART Status.
+ * This method will attempt to:
+ * 1. Retrieve the MRIE value of the SAS drive
+ * 2. Depending on the MRIE value, we will either submit a REQUEST SENSE
+ *    command or a LOG SENSE command.
+ * 3. Return the asc/ascq values to the caller.
+ * 
+ * Input *asc should be a pointer to a two byte unsigned number.
+ * Input *ascq should be a pointer to a two bytes unsigned number.
+ * The maximum *sd_name strlen is (_MAX_SD_NAME_STR_LEN - 1), assuming
+ * the caller did the check.
+ * Return LSM_ERR_NO_MEMORY or LSM_ERR_NO_SUPPORT or LSM_ERR_LIB_BUG or
+ * LSM_ERR_NOT_FOUND_DISK.
+ */
+int _sg_sas_smart_status(char *err_msg, int fd, int32_t *smart_status)
+{
+    int rc = LSM_ERR_OK;
+    uint8_t informational_exceptions_mode_page[_SG_T10_SPC_MODE_SENSE_MAX_LEN];
+    uint8_t informational_exceptions_log_page[_T10_SPC_LOG_SENSE_MAX_LEN];
+    uint8_t asc = 0;
+    uint8_t requested_sense[_T10_SPC_SENSE_DATA_MAX_LENGTH];
+    struct _sg_t10_sense_fixed *sense_fixed = NULL;
+    struct _sg_t10_informational_exceptions_mode_page_0_hdr *ie_mode_hdr = NULL;
+    struct _sg_t10_informational_exceptions_general_log_hdr *ie_log_hdr = NULL;
+
+    _good(_sg_io_mode_sense(err_msg, fd, INFO_EXCEPTION_CONTROL_PAGE, 0,
+                            informational_exceptions_mode_page),
+          rc, out);
+    ie_mode_hdr = (struct _sg_t10_informational_exceptions_mode_page_0_hdr *)
+                  informational_exceptions_mode_page;
+    
+    if (ie_mode_hdr->mrie == MRIE_REPORT_INFO_EXCEPTION_ON_REQUEST) {
+        _good(_sg_request_sense(err_msg, fd, requested_sense), rc, out);
+        sense_fixed = (struct _sg_t10_sense_fixed *) requested_sense;
+        asc = sense_fixed->asc;
+    } else {
+        _good(_sg_log_sense(err_msg, fd, _T10_SPC_INFO_EXCEPTION_PAGE_CODE, 0,
+                            informational_exceptions_log_page), rc, out);
+        // SPC5 rev07 - Table 349 - Informational Exceptions General log parameter
+        ie_log_hdr = (struct _sg_t10_informational_exceptions_general_log_hdr *)        
+                     informational_exceptions_log_page;
+        asc = ie_log_hdr->asc;
+    }
+
+    *smart_status = _sg_smart_interpret_asc(asc);
+
+out:
+    return rc;
+}
+
+/*
+ * Query a SATA drive attached via SAS to get its SMART Status.
+ * 
+ * Input *smart_status should be a pointer to a one byte unsigned number.
```

Comment:
@tasleson Good catch...I thrashed a little bit on return values while I was working on this.